Add initial parsec-tool implementation

joechrisellis opened the pull request:
Hi guys, this is the initial parsec-tool implementation from issue 202 in [parallaxsecond/parsec](https://github.com/parallaxsecond/parsec).

Any and all comments welcome and appreciated! :)

ionut-arm commented on src/command.rs:
```diff
@@ -0,0 +1,105 @@
+/// Gets the list of availble opcodes for a given provider and outputs them.
+fn list_opcodes(
+    provider: &str,
+    _command_line_context: &CommandLineContext,
```
I'd say, for all these functions, if you don't need this argument you shouldn't really pass it in/request it

joechrisellis replied:
Sure, will remove.

joechrisellis replied:
Actually, I'm not so sure about this -- in the future it's likely that we'll want to show more/less information depending on the verbosity, which is encoded in the command line context. I think this is a good thing to keep around, even if we're not using it right now.

ionut-arm replied:
Makes sense

ionut-arm commented on src/command.rs:
```diff
@@ -0,0 +1,105 @@
+        print_colored!(Cyan, "0x{:02x} ({})", provider.id as u32, provider.id);
+        println!("]");
+
+        print_colored!(Yellow, "Description");
```
If you keep this system for colouring output, you should probably write down somewhere what each colour represents (for example, why the details here are yellow - I would've assumed yellow is more of a warning colour).

ionut-arm replied:
I'm not saying yellow is wrong, btw, just that it should be clear what the colours are used for

joechrisellis replied:
That's a good idea -- so far, the format is:

* Yellow to represent field names -- i.e. 'description', 'uuid'.
* Cyan for titles.
* Blue for information.
* Red for errors (and warnings?).
* Green for success messages.

ionut-arm replied:
If we agree on a list like this, would it make sense to make separate macros for each of these to handle both the colour and the printing inside the macro? So for example, instead of
```rust
print_colored!(Yellow, "Description");
println!(": {}", provider.description);
```
you'd get the same result via
```rust
print_field!("Description", "{}", provider.description);
```
Apologies if this is a bit over the top 😅

ionut-arm commented on src/output.rs:
```diff
@@ -0,0 +1,112 @@
+    };
+}
+
+/// Write a success message to stderr.
+#[macro_export]
+macro_rules! e_success {
+    ($($arg:tt)*) => {
+        eprint!("{}", colorize!(Green, Stderr, "[SUCCESS] "));
+        eprintln!($($arg)*);
+    };
+}
```
I'm somewhat confused by this - why write a success message to stderr?

joechrisellis replied:
AFAIK stderr is not just used for error messages, but diagnostic messages that aren't part of a program's 'functional output' too. It's definitely possible to have a success message that's just a diagnostic message, which is why the option is available. 😄

ionut-arm replied:
That's definitely good to know, thanks! A very unfortunate naming choice for the stream, though 😄

joechrisellis commented on src/output.rs:
```diff
@@ -0,0 +1,112 @@
+// Copyright 2020 Contributors to the Parsec project.
```
I took a look around for some libraries to handle logging/style but I couldn't find anything that was as flexible as `ansi_term`, which is what I've opted for here. Logging felt a little bit extreme, since we're not really 'logging' as much as we are just writing to std{out,err}. Happy to change this if you think it's a better long term solution though!

ionut-arm replied:
No, this makes sense, it's not really logging in the end :) I think I was mostly bothered by the thought of having to set the colour for each line that we write, feels a bit extreme.

joechrisellis commented on src/cli.rs:
```diff
@@ -0,0 +1,65 @@
+#[derive(Debug)]
+pub enum Action {
+    ListOpcodes(String),
+    ListProviders(),
+    Ping(),
```
Nope -- removed them. :)

hug-dev commented on src/main.rs:
```diff
@@ -33,6 +33,26 @@
+use crate::command::dispatch_command;
+use anyhow::{Context, Result};
+
+fn run() -> Result<bool> {
```
Is the `bool` result actually used? If not, this function can maybe return `Result<()>`.

joechrisellis replied:
No, not yet. The idea was going to be:

* Return `Err` on complete failure.
* Return `Ok(true)` on success with partial failure.
* Return `Ok(false)` on complete success.

But since we aren't using this right now, I agree -- better to just remove it.

hug-dev commented on src/cli.rs:
```diff
@@ -0,0 +1,65 @@
+                Arg::with_name("provider")
+                    .short('p')
+                    .takes_value(true)
+                    .required(true)
```
Could the `p` argument be not required and if not passed, the `list_opcodes` command would execute on the default provider (the first one returned by `list_providers`)?

joechrisellis replied:
Good idea, will change this.

hug-dev commented on src/cli.rs:
```diff
@@ -0,0 +1,65 @@
+        } else {
+            unreachable!()
+        },
+        app_name: match matches.value_of("appname") {
```
Is `appname` one of the possible option? I don't see it in `matches`

hug-dev commented on src/cli.rs:
```diff
@@ -0,0 +1,65 @@
+use clap::{App, AppSettings, Arg};
+
+#[derive(Debug)]
+pub enum Action {
```
In the future, `Action` should probably be replaced by [`NativeOperation`](https://docs.rs/parsec-interface/0.19.0/parsec_interface/operations/enum.NativeOperation.html) to not have to duplicate similar information. This module would create the correct instance of `NativeOperation` from the command entered by the user.
Having an automatic translation (as a procedural macro maybe?) of each of the `Operation` into command line arguments for `clap` would be phenomenal.

ionut-arm replied:
> Having an automatic translation (as a procedural macro maybe?) of each of the `Operation` into command line arguments for `clap` would be phenomenal.

I don't think that's a good idea for some of the operations at least, especially the ones that take very rich data structures. Imagine having to write the whole `KeyAttributes` structure on the command line. It's probably more sensible to expect stuff like that to be written in a TOML file or something similar, and we parse it from there.

Hence, I'm not sure just automating the whole translation process makes that much sense, unless we can specify, for example, which parameters of a command should be taken from files, and how the file should be converted to the parameter in the first place. Even for "very simple" data like a file that you want to encrypt, it's probably more reasonable to let the user decide if they want to pipe the contents through the CLI or have an option to select the file based on its path (and potentially write a signature for it in a different file, instead of just blindly writing stuff to stdout).

hug-dev replied:
That's true, I did not think about `KeyAttributes`, that would be very painful to write all down. Maybe more generally the CLI should limit itself to the highest abstraction level of the client? When we have features for parameters discovery and simple abstraction over complicated operations, then the CLI can have commands over that.

hug-dev commented on src/output.rs:
```diff
@@ -0,0 +1,123 @@
+// - Cyan should be used for section titles.
+// - Yellow should represent field names -- i.e. UUIDs, versions, etc.
+//
+// By sticking to this guideline we ensure that the CLI stays consistent.
```
Agree with Ionut that it would be nice to have macros mapping with each one of those colours to be consistent everywhere. Also, it would maybe be nice to not have the `macro_export` but let the module where they are needed import them directly like static functions.

hug-dev commented on src/output.rs:
```diff
@@ -0,0 +1,123 @@
+
+/// Write an info message to stdout.
+#[macro_export]
+macro_rules! info {
```
Is the verbosity level argument coming into play for the `info`, `warn` and `err` macros? Would it be possible to use a logger like `env_var` directly? I think it also handles colors.

ionut-arm commented on src/main.rs:
```diff
@@ -33,6 +33,26 @@
 //! Parsec Tool: a tool to communicate with Parsec from the command-line
 
+#[macro_use]
+mod output;
+
+mod cli;
+mod command;
+mod common;
+mod error;
```
Do you intend that documentation for users of the tool or for its developers?

hug-dev replied:
For developpers only, the library crate exposed should probably not be used.

joechrisellis commented on src/main.rs:
```diff
@@ -33,6 +33,26 @@
 
 //! Parsec Tool: a tool to communicate with Parsec from the command-line
 
+#[macro_use]
```
I don't think so -- the purpose of this line is to make explicitly make the `output` module available in the crate. If we want to use its macros, I think `macro_use` is necessary. 🙁

hug-dev replied:
What about something like:

```rust
mod output;
use output::e_err;
```
?

joechrisellis commented:
Hi guys. I've addressed your comments as well as done some further work on this. The initial implementation I have now is quite different to what is shown in this patch. Do you want me to close this PR and open a new one or would you rather I just push the results here?

ionut-arm commented:
> 
> 
> Hi guys. I've addressed your comments as well as done some further work on this. The initial implementation I have now is quite different to what is shown in this patch. Do you want me to close this PR and open a new one or would you rather I just push the new stuff here?

Push it over this one, if someone asks why some things are the way they are, we can point them to this PR and it'll have the full convo.

hug-dev commented:
Is everything changing in the new version? Would it make sense for you to first apply the changes requested here, merge this PR and then open a new one with the new changes?

Maybe it does not make sense if the comments we made are not relevent anymore.

joechrisellis commented:
@hug-dev a lot has changed -- I think I will just re-push the code here just to keep the conversation going. 🙂

ionut-arm commented on src/subcommands/mod.rs:
```diff
@@ -0,0 +1,54 @@
+
+/// Runs the appropriate command given `matches`, which corresponds to the command-line invocation
+/// from the user.
+pub fn dispatch_subcommand(matches: ParsecToolApp) -> Result<(), ParsecToolError> {
```
Similar to the previous comment, this can be implemented as a method on `Subcommand`, or maybe even on `ParsecToolApp`, depending on how you want to call it in `main()`. 

We've generally avoided having "hanging" functions if we can implement them on some type where it makes sense for them to be attached. Sometimes that's not possible, though, either because the types are external so we literally cannot implement on them, or because we need multiple implementations for different things, so we then use functions.

ionut-arm commented on src/subcommands/mod.rs:
```diff
@@ -0,0 +1,54 @@
+/// - They implement `run`, which executes the subcommand.
+pub trait ParsecToolSubcommand: StructOpt + TryInto<NativeOperation> {
+    /// Run the subcommand.
+    fn run(self, matches: ParsecToolApp) -> Result<(), ParsecToolError>;
```
I'd probably split that `ParsecToolApp` into `Subcommand` and `Parameters` or `Options` or something like that, where you group all the other parts. And then in this method you pass a reference to that other struct, it feels fishy to give to a subcommand ownership over the structure that presumably holds it? I'm assuming this works because `Subcommand` has `Copy` on it

joechrisellis commented on src/subcommands/list_providers.rs:
```diff
@@ -0,0 +1,70 @@
+#[structopt(name = "list_providers")]
+pub struct ListProvidersSubcommand {}
+
+impl TryInto<NativeOperation> for ListProvidersSubcommand {
```
Done, good spot!

joechrisellis commented on src/subcommands/common.rs:
```diff
@@ -0,0 +1,27 @@
+/// effectively a `FromStr` implementation for AuthenticationData. Passing in `None` will return
+/// AuthenticationData::None. Passing in `Some(s)` will give you an app identity whose secret is
+/// built from the string `s`.
+pub fn make_authentication_data(auth_secret: Option<String>) -> AuthenticationData {
```
Quite right! Done. :)

joechrisellis commented on src/main.rs:
```diff
@@ -1,38 +1,25 @@
+
+fn run() -> Result<()> {
+    let matches = cli::ParsecToolApp::from_args();
+    dispatch_subcommand(matches).context("Executing subcommand failed.")?;
+    Ok(())
```
Good spot! I think I'll leave this as-is, though -- to me, the more explicit use of `Ok(())` is more readable. If you have strong opinions in this I'm happy to change though. 😄

hug-dev commented on src/lib.rs:
```diff
@@ -0,0 +1,43 @@
+pub mod cli;
+pub mod common;
+pub mod error;
+pub mod subcommands;
```
I like the tree structure you choosed, with `subcommands` as a module!

hug-dev commented on src/subcommands/list_opcodes.rs:
```diff
@@ -0,0 +1,62 @@
+    provider_opts: ProviderOpts,
+}
+
+impl TryFrom<ListOpcodesSubcommand> for NativeOperation {
+    type Error = ParsecToolError;
+
+    fn try_from(list_opcodes_subcommand: ListOpcodesSubcommand) -> Result<Self, Self::Error> {
+        // Trivially converted to a `NativeOperation`.
+        Ok(NativeOperation::ListOpcodes(list_opcodes::Operation {
+            provider_id: ProviderID::try_from(list_opcodes_subcommand.provider_opts.provider)?,
+        }))
+    }
+}
```
Nice! It's a very good idea to use the `OperationClient` here. Also each command will make its own choice of either using the basic client or the operation client 👌

hug-dev commented on src/cli/mod.rs:
```diff
@@ -0,0 +1,49 @@
+pub struct ParsecToolApp {
+    /// How verbose should we be?
+    #[structopt(short = "v", multiple = true, default_value = "0")]
+    pub verbosity: u8,
```
This does not seem to be used in the PR, are you going to implement it later?

joechrisellis replied:
Yes, I was thinking that verbosity would be useful to control the level of output for some of the more complex operations in the future. It's unused at the moment though. 🙂

hug-dev commented on src/util/output.rs:
```diff
@@ -0,0 +1,157 @@
+
+/// Write an info message to stdout.
+#[macro_export]
+macro_rules! info {
+    ($($arg:tt)*) => {
+        $crate::print_colored!(Blue, "[INFO] ");
+        println!($($arg)*);
+    };
+}
```
I think we should only keep one version of the macros, choosing only to emit logs to stdout or stderr. I would say that here we can simply log everything to stderr. If we want to have the choice, it might be best to have a configuration option to choose between stdout or stderr but for simplicity I think it's best to just log everything to stderr.

hug-dev replied:
On that note, maybe we can delete those macros and directly use the one from `log` with the `env_logger` implementor? The `verbosity` flag could set the log level and I am pretty sure that this crate handles colouring as well 💯 It's also customizable to make it look nice for the command line.

joechrisellis replied:
Okay, I'll get rid of the stdout stuff!

I think using the `env_logger` module is a good idea, but I think I'll implement this in a future commit if that's alright with you? It might require some further change throughout this repo. 😛

hug-dev replied:
Sure that's fine!

joechrisellis commented on src/cli/mod.rs:
```diff
@@ -0,0 +1,49 @@
+
+/// Struct representing the command-line interface of parsec-tool.
+#[derive(Debug, StructOpt)]
+#[structopt(name=PROJECT_NAME, about=PROJECT_DESC, author=PROJECT_AUTHOR, version=PROJECT_VERSION)]
```
It does, but doesn't seem to do it for the authors, so it makes sense to just leave them all there IMO. 🙂

joechrisellis commented on src/lib.rs:
```diff
@@ -0,0 +1,43 @@
+#![deny(
+    nonstandard_style,
+    const_err,
+    // dead_code,
```
That's a debugging artifact, my bad.

joechrisellis commented on src/cli/mod.rs:
```diff
@@ -0,0 +1,49 @@
+    /// effectively a `FromStr` implementation for AuthenticationData. Passing in `None` will return
+    /// AuthenticationData::None. Passing in `Some(s)` will give you an app identity whose secret is
+    /// built from the string `s`.
+    pub fn make_authentication_data(&self) -> AuthenticationData {
```
Good idea, done.

joechrisellis commented on src/error.rs:
```diff
@@ -0,0 +1,24 @@
+//! Error definitions/handling.
+
+use parsec_interface::operations::NativeResult;
+use thiserror::Error;
```
It is! 😄

hug-dev commented on src/cli/mod.rs:
```diff
@@ -0,0 +1,49 @@
+
+impl ParsecToolApp {
+    /// Run the requested subcommand.
+    pub fn dispatch_subcommand(&self) -> Result<(), ParsecToolError> {
```
Nitpicking here but in a future version, if that is ok, `Subcommand` itself could implement the `ParsecToolSubcommand` trait so that you could call `run` on it directly from `main` without needing this method.
